diff --git a/compiler/rustc_interface/src/util.rs b/compiler/rustc_interface/src/util.rs
index 23b152bd57240fb9f0eb995170476388d2b2d999..619a2db6b0415b0c22abd486d1a1afc67724be81 100644
--- a/compiler/rustc_interface/src/util.rs
+++ b/compiler/rustc_interface/src/util.rs
@@ -151,6 +151,7 @@ fn init_stack_size(early_dcx: &EarlyDiagCtxt) -> usize {
     })
 }
 
+#[cfg(not(all(target_family = "wasm", not(target_feature = "atomics"))))]
 fn run_in_thread_with_globals<F: FnOnce(CurrentGcx) -> R + Send, R: Send>(
     thread_stack_size: usize,
     edition: Edition,
@@ -190,6 +191,21 @@ fn run_in_thread_with_globals<F: FnOnce(CurrentGcx) -> R + Send, R: Send>(
     })
 }
 
+/// WebAssembly without threads can't spawn one, so run on the current thread.
+/// Its stack size is then set when linking, not here. (Patched in by rust-js.)
+#[cfg(all(target_family = "wasm", not(target_feature = "atomics")))]
+fn run_in_thread_with_globals<F: FnOnce(CurrentGcx) -> R + Send, R: Send>(
+    _thread_stack_size: usize,
+    edition: Edition,
+    sm_inputs: SourceMapInputs,
+    extra_symbols: &[&'static str],
+    f: F,
+) -> R {
+    rustc_span::create_session_globals_then(edition, extra_symbols, Some(sm_inputs), || {
+        f(CurrentGcx::new())
+    })
+}
+
 pub(crate) fn run_in_thread_pool_with_globals<F: FnOnce(CurrentGcx) -> R + Send, R: Send>(
     thread_builder_diag: &EarlyDiagCtxt,
     edition: Edition,
